R/data.table.R

```diff
@@ -2878,22 +2880,11 @@ setDT = function(x, keep.rownames=FALSE, key=NULL, check.names=FALSE) {
     x = null.data.table()
   } else if (is.list(x)) {
     # copied from as.data.table.list - except removed the copy
-    for (i in seq_along(x)) {
-      if (is.null(x[[i]])) next   # allow NULL columns to be created by setDT(list) even though they are not really allowed
-                                  # many operations still work in the presence of NULL columns and it might be convenient
-                                  # e.g. in package eplusr which calls setDT on a list when parsing JSON. Operations which
-                                  # fail for NULL columns will give helpful error at that point, #3480 and #3471
-      if (inherits(x[[i]], "POSIXlt")) stopf("Column %d is of POSIXlt type. Please convert it to POSIXct using as.POSIXct and run setDT again. We do not recommend use of POSIXlt at all because it uses 40 bytes to store one date.", i)
-    }
-    n = vapply_1i(x, length)
```

Comment:
If a R 3.6.2 dependency is OK, `n = lengths(x)` provides a 4 second boost (4.1 s -> 0.1s) on my computer on your 1e7 list compared to `n = vapply(x, length, 1L)`. The `setDT(dt)` takes 35 s on my computer, so it still would be pretty slow.

But if C code is appropriate, `rbindlist.c` and `uniqlist.c` have similar concepts about lengths being equal, especially in `rbindlist.c`. So it might be a good investment to refactor to have C code determine if all elements of a list have equal lengths so it could be re-used.

Reply:
3.6.2 is waaaaay past our current dep so it is off the table for the foreseeable future I'm afraid :/

agree about sharing code where possible. I'll look into it. might mark it as a follow-up task though if it requires some refactoring

Reply:
I misspoke and `lengths` is actually from 3.2.

Reply:
Good point, I can make the change after #5838.

Reply:
`lengths()` is already in the C portion, resolving this